frontend/lib/swr-cache/index.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * withCache is a function that enables caching the result of any async operation using a stale while revalidate strategy.
+ * The function takes a set of options and returns a NextApiHandler that you can use in your API routes to perform
+ * background revalidation of the operation that you're caching. The handler also exposes a get function that you can
+ * use to get the cached value (or a fresh value on cache misses).
+ * Whenever you call the get function and the data is stale, a background revalidation is automatically requested.
+ *
+ * Usage:
+ *
+ * 1) Define an operation endpoint in your API routes:
+ *    e.g. `export default withCache({ ... })`
+ * 2) Import the operation endpoint where you need to get the value:
+ *    e.g. `import ProductList from '@pages/api/product-list';`
+ * 3) Use the get function to get the cached value:
+ *   e.g. `const productList = await ProductList.get({ ... })`
+ */
+
+import { APP_ORIGIN } from '@config/env';
+import { log as defaultLog, Logger, nullLog } from '@lib/logger';
+import type { NextApiHandler } from 'next';
+import { z } from 'zod';
+import { getCache } from './adapters';
+import {
+   createCacheEntry,
+   createCacheKey,
+   isStale,
+   isValidCacheEntry,
+   sleep,
+} from './utils';
+
+interface CacheOptions<
+   VariablesSchema extends z.ZodTypeAny,
+   ValueSchema extends z.ZodTypeAny
+> {
+   endpoint: string;
+   variablesSchema: VariablesSchema;
+   valueSchema: ValueSchema;
+   getFreshValue: (
+      variables: z.infer<VariablesSchema>
+   ) => Promise<z.infer<ValueSchema>>;
+   ttl?: number;
+   staleWhileRevalidate?: number;
+   log?: boolean;
+}
+
+type NextApiHandlerWithProps<
+   Variables = unknown,
+   Value = unknown
+> = NextApiHandler & {
+   get: (variables: Variables) => Promise<Value>;
+   revalidate: (variables: Variables) => Promise<void>;
+};
+
+export const withCache = <
+   VariablesSchema extends z.ZodTypeAny,
+   ValueSchema extends z.ZodTypeAny
+>({
+   endpoint,
+   variablesSchema,
+   getFreshValue,
+   ttl,
+   staleWhileRevalidate,
+   log = false,
+}: CacheOptions<VariablesSchema, ValueSchema>): NextApiHandlerWithProps<
+   z.infer<VariablesSchema>,
+   z.infer<ValueSchema>
+> => {
+   const logger: Logger = log ? defaultLog : nullLog;
+   const cache = getCache();
+
+   const requestRevalidation = async (variables: z.infer<VariablesSchema>) => {
+      fetch(`${APP_ORIGIN}/${endpoint}`, {
+         method: 'POST',
+         body: JSON.stringify(variables),
+      }).catch((error) => {
+         logger.error(
+            `[cache > start background revalidation]: unable to trigger revalidation\n${error}`
+         );
+      });
+      // We add a small delay to ensure the revalidation is triggered
+      await sleep(50);
```

Comment:
I don't understand why revalidation wouldn't be triggered without this sleep. Could you explain it to me?

Reply:
Because since we're not awaiting the fetch request, the response might be sent (hence the process terminated) before node has even the chance to start the network call. Adding a small delay helps make sure that the network request gets initiated. We can play with the delay a bit, but I fill like it's not necessary to optimize this for the time being since it should be sufficiently fast.

Reply:
Makes sense, thanks! I agree that we do not need to optimize this.